library/Booster/Pattern/Base.hs

```diff
@@ -153,37 +161,49 @@ instance Recursive Term where
 getAttributes :: Term -> TermAttributes
 getAttributes (Term a _) = a
 
-kmapUnitSymbol, kmapConcatSymbol, kmapElementSymbol :: KMapDefinition -> Symbol
-kmapUnitSymbol def =
+unitSymbol, concatSymbol :: KCollectionMetadata -> Symbol
+unitSymbol def =
     Symbol
-        { name = def.symbolNames.unitSymbolName
+        { name = symbolNames.unitSymbolName
         , sortVars = []
         , argSorts = []
-        , resultSort = SortApp def.mapSortName []
+        , resultSort = SortApp collectionSort []
         , attributes =
             SymbolAttributes
                 { symbolType = TotalFunction
                 , isIdem = IsNotIdem
                 , isAssoc = IsNotAssoc
                 , isMacroOrAlias = IsNotMacroOrAlias
-                , isKMapSymbol = Just def
+                , collectionMetadata = Just def
                 }
         }
-kmapConcatSymbol def =
+  where
+    (symbolNames, collectionSort) =
+        case def of
+            KMapMeta mapDef -> (mapDef.symbolNames, mapDef.mapSortName)
+            KListMeta listDef -> (listDef.symbolNames, listDef.listSortName)
+concatSymbol def =
     Symbol
-        { name = def.symbolNames.concatSymbolName
+        { name = symbolNames.concatSymbolName
         , sortVars = []
-        , argSorts = [SortApp def.mapSortName [], SortApp def.mapSortName []]
-        , resultSort = SortApp def.mapSortName []
+        , argSorts = [SortApp collectionSort [], SortApp collectionSort []]
+        , resultSort = SortApp collectionSort []
         , attributes =
             SymbolAttributes
-                { symbolType = PartialFunction
+                { symbolType
                 , isIdem = IsNotIdem
                 , isAssoc = IsAssoc
                 , isMacroOrAlias = IsNotMacroOrAlias
-                , isKMapSymbol = Just def
+                , collectionMetadata = Just def
                 }
         }
+  where
+    (symbolNames, collectionSort, symbolType) =
+        case def of
+            KMapMeta mapDef -> (mapDef.symbolNames, mapDef.mapSortName, PartialFunction)
+            KListMeta listDef -> (listDef.symbolNames, listDef.listSortName, TotalFunction)
+
+kmapElementSymbol :: KMapDefinition -> Symbol
```

Comment:
why not merge `kmapElementSymbol` with `klistElementSymbol` like the two other cases above?

Reply:
I thought this was safer because the returned symbol has a different type and arity for the two different cases.